### ISSUE TITLE
Connection terminated unexpectedly (Postgresql / AWS Lambda)

### ISSUE THREAD

**Issue by jamesdixon** [knex/knex#3636]
# Environment

Knex version: 0.20.8
Database + version: Postgres 9.5.14
OS: AWS Lambda - Node 12.x

Recently, I've been seeing upticks in a `connection terminated unexpectedly` error. I can't find a discernible pattern at this point but unlike #3523, it doesn't seem to occur after long periods of activity. I see instances that are minutes apart in some cases. 

I added some logging to the Knex `afterCreate` method and am seeing two errors:

1. The `connection terminated unexpectedly` error
2. And `Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed` - this seems to happen on the 1st or second invocation after seeing the `connection terminated unexpectedly` error.

I've tried a number of configurations with Knex pooling:

1. `{ pool: { min: 1, max: 1 } }`
2. `{ pool: { min: 0  }`
3. `{ pool: { min: 1 } }`

I've also tried setting `{ pool: { idleTimeoutMillis: 300000 } }` as per #3523.

Additional information:
* I'm defining my Knex instance outside of my AWS handlers, so the pool should be reused per container.
* I'm using the following for my `afterCreate` function -- I'm not sure if there's a better way to get at the underlying error?

```js
function afterCreate(connection, done) {
  connection.on('error', (err) => {
    console.log('POSTGRES CONNECTION ERROR');
    console.error(err)
  });
  done()
}
```
Appreciate any help! This is driving me nuts :)

**UPDATE:** the only change that I've made is migrating the lambda function to Node 12 from Node 8 as AWS is removing support for 8. Unfortunately, it doesn't look like I can go back. My other functions in production are on Node 8 and not seeing any issues 🤔🤔🤔

**ISSUE FOUND:** for whatever reason, reverting back to Node 8.x fixes the issue. Still not certain on the underlying cause.

**Comment by elhigu** [comment]
have you tried hooking some logging to:

knex.on('query', ...) , knex.on('query-success', ...), knex.on('query-error', ...), knex.on('error', ...)

Also you could wire tarn.js hooks to output what pool is doing when errors happen. Tarn instance is found from `knex.client.pool`

**Comment by jamesdixon** [comment]
@elhigu I have not, but I'll give it a shot. Odd that this seems to only occur with Node 10/12. Thanks for the tips!

**Comment by FreddyJD** [comment]
I'm having the same issue with MySQL. Same behavior

**Comment by jamesdixon** [comment]
@FreddyJD ah, good to know that I'm not the only one.

As I mentioned above, reverting back to Node 8.x solves the issue. That said, I've had to resort to running a custom Node 8 runtime on Lambda as it's officially been deprecated so functions can no longer be created or updated to Node 8.

Have you found that Node 8 resolves the problem?

**Comment by mouhannad-sh** [comment]
I'm having the same issue too with Postgres

**Comment by FreddyJD** [comment]
@jamesdixon We've read about setting ```   context.callbackWaitsForEmptyEventLoop = false; ```, so the function shutdowns when the event loop is finished. I'll have to wait until we release it.

I found the problem in Node10.X

**Comment by mouhannad-sh** [comment]
@FreddyJD in production we are using
```
context.callbackWaitsForEmptyEventLoop = false;
```
and the issue is still there
I just made a new release where I work and removed that line, we'll have to wait for few hours to monitor as the issue happens randomly

**Comment by jamesdixon** [comment]
@FreddyJD @mouhannad-sh our functions also were running `context.callbackWaitsForEmptyEventLoop = false;` in production but as mentioned, the issue didn't pop up until Node 10+. I guess the interesting thing is that this happening for both MySQL and Postgres. More likely that it's a Knex-related issue then?

**Comment by mouhannad-sh** [comment]
it probably is, we've spotted the same error happening again after removing that line

**Comment by 592da** [comment]
any solution? urgent for us. thanks

**Comment by jamesdixon** [comment]
@592da we ended up using a custom Node 8 runtime, which has worked for us. Not ideal but we didn't have the time or resources to dig deep into this issue -- we just needed things to work :)

**Comment by briandamaged** [comment]
Howdy!  I'll try to investigate this in the next day or so.  (Sry -- I'm currently assisting w/ a large merge request.  I want to get that resolved before making additional changes to the code)

**Comment by briandamaged** [comment]
@jamesdixon : Can you plz provide the full stack trace for the `"Connection terminated unexpectedly"` that you are observing?

Based upon my initial investigation, it looks like the `"pg"` package is most likely throwing this error (as opposed to `"knex"` throwing it).  The error seems to occur when:

1. Something has closed the low-level connection w/ the PG server, but
2. The `"pg"` library was not expecting this connection to be ended yet.  (ie: there's a higher-level `Client` class that is supposed to initiate the connection termination.  It throws the `"Connection terminated unexpectedly"` error when it notices that _something else_ terminated the connection)

As far as I can tell, `"knex"` only interacts w/ the `"pg"` library's `Client` class.  This suggests (but does not prove) that something else might be killing the lower-level connection.

Out of curiosity: did your project's `package-lock.json` change when you upgraded from `node 8` -> `node 12`?  If so: did the version of the `"pg"` library change as well?

**Comment by jamesdixon** [comment]
@briandamaged  you're correct on the stack trace pointing to the `pg` client. Seems like someone else opened up a similar issue here: https://github.com/brianc/node-postgres/issues/2112

```
Error: select * from "appointment" where "id" = $1 limit $2 - Connection terminated unexpectedly
at Connection.con.once (/var/task/node_modules/pg/lib/client.js line 255 col 8)
: new Error('Connection terminated unexpectedly')
at Object.onceWrapper (events.js line 313 col 29)
at emitNone (events.js line 106 col 12)
at Connection.emit (events.js line 208 col 6)
at Socket.<anonymous> (/var/task/node_modules/pg/lib/connection.js line 78 col 9)
self.emit('end')
at emitOne (events.js line 116 col 12)
at Socket.emit (events.js line 211 col 6)
at TCP._handle.close [as _onclose] (net.js line 561 col 11)
```

Let me check on your question about `package-lock.json`.

**Comment by briandamaged** [comment]
@FreddyJD 

> I'm having the same behavior with MySQL.

Can you please provide the full error message / stacktrace?  I suspect that you are observing the similarly-worded `"Connection ENDED unexpectedly"` error, which might be pointing to a separate problem.

**Comment by briandamaged** [comment]
@FreddyJD 

(To clarify: the `"Connection terminated unexpectedly"` error is not produced by Knex.  However, Knex does produce a `'Connection ended unexpectedly'` error via a separate control path)

**Comment by briandamaged** [comment]
On the `"pg"` side: it looks like there might be a bug in the `"pg"` library itself.  Consequently, the `pg` library does not realize that a `Client`'s connection has ended until after it tries to use it again.  This then produces the `"Connection terminated unexpectedly"` error.

Interestingly: once this error has been thrown, the problem gets "fixed" on the `pg` side.  Afterwards, `knex` detects that the client's connection has ended, emits a `"Connection ended unexpectedly"` warning, and fixes its connection pool.

**Comment by briandamaged** [comment]
Another quick update on this front:

Currently, it's looking like both `"pg"` and `"knex"` have the same bug within their respective pool implementations.  Specifically: neither pool implementation appears to be re-validating the connection before handing it off to the caller.  This is providing an opportunity for the connections to the closed silently while they are sitting idle within the pool.

So, some near-term steps:

1. The `"pg"` team is currently trying to verify a proposed fix.  If that works, then
2. They'll refactor the fix and make some of the details available in a public-facing API.  (ie: rather than having to check the internal details of some deeply-nested objects)
3. Afterwards, `"knex"` can be upgraded to leverage this public-facing API and fix the issue within its PG connection pooling.

**Comment by briandamaged** [comment]
cc: @brianc   ☝️

**Comment by elhigu** [comment]
I had been suggesting that the problem is actually some proxy server in AWS lambda network infra, which would be acting like the connection is just fine until some real query is sent through it. Maybe AWS postgres implementation just work different 🤷‍♂

Anyways it really would be nice if pg driver is able to recognize the broken connection internally and emit an error or end event for that client instance.  Then knex could just subscribe to that and set everything so that broken connection would be discarded from pool during the validate.

**Comment by briandamaged** [comment]
@elhigu : Agreed -- there's a good chance that it's a piece of network infrastructure that is discarding the connection silently.  The hope is that there is some way to detect this condition.  If so, then the pool implementations can verify that the connections are still valid before handing them off to the caller.

**Comment by jamesdixon** [comment]
Do you guys have any thoughts about why this would be present with Node 10 / 12 but not Node 8?

**Comment by kibertoad** [comment]
First thing that comes to mind is changes in EventEmitter logic between Node 8 and Node 10.

**Comment by 592da** [comment]
I tried to use nodejs version 8 custom runtime for lambda, and still issue occurs.
@kibertoad @jamesdixon @briandamaged

**Comment by michaelseibt** [comment]
> There's a good chance that it's a piece of network infrastructure that is discarding the connection silently. The hope is that there is some way to detect this condition. If so, then the pool implementations can verify that the connections are still valid before handing them off to the caller.

I think this is the gist of it. AWS freezes a Lambda for later usage, which includes memory pointers (variables and such) that have been created in a global-ish scope ([this post on Medium summarizes it)](https://medium.com/@feliperohdee/aws-lambda-reusing-database-connections-16343eb74f9f).
I would assume this includes pooled clients, which never get a chance to react to closing Sockets.

From the [AWS documentation](https://docs.amazonaws.cn/en_us/lambda/latest/dg/runtimes-context.html):

> After a Lambda function is executed, AWS Lambda maintains the execution context for some time in anticipation of another Lambda function invocation. In effect, the service freezes the execution context after a Lambda function completes, and thaws the context for reuse, if AWS Lambda chooses to reuse the context when the Lambda function is invoked again. This execution context reuse approach has the following implications: Objects declared outside of the function's handler method remain initialized, providing additional optimization when the function is invoked again. For example, if your Lambda function establishes a database connection, instead of reestablishing the connection, the original connection is used in subsequent invocations. We suggest adding logic in your code to check if a connection exists before creating one.

**Comment by elhigu** [comment]
@jamesdixon this has been around as long as AWS lambda with pg has existed. 

> AWS freezes a Lambda for later usage, which includes memory pointers (variables and such) that have been created in a global-ish scope (this post on Medium summarizes it).
I would assume this includes pooled clients, which never get a chance to react to closing Sockets.

I don't think that is enough to make knex / pg act like that. Though this scenario is easy to test. Adding for example 1 second delay in start of lambda function before getting connection from pool would give knex/pg/node eventloop enough time to notice that connection was broken before it is fetched from pool.

> We suggest adding logic in your code to check if a connection exists before creating one.

Knex has this kind of logic builtin.

**Comment by briandamaged** [comment]
> Knex has this kind of logic builtin.

@elhigu : It has it built-in, but the base implementation still relies upon certain Events being fired.  These events attach a `__knex__disposed` field to the `connection` object to indicate that the connection has been closed.  This field is then checked by the `pool` before handing off the `connection`.

So, I believe this is the reason that some end-users are still seeing this issue in `knex`.  If the Events are not fired/handled quickly enough, then a connection can slip through the cracks.

Fortunately, `knex` also provides a hook so that each `Dialect` can add its own `connection` validation logic.  For `Client_PG`, this method currently just returns `true`.  If the `pg` team adds a public method for checking the state of a `Client`, then we can update the `Client_PG`'s hook to query this state.

(Sry if that's confusing.  Both the `knex` and `pg` projects use the term `Client` to refer to slightly-different things.  A `pg` `Client` is a `knex` `connection`. )

Alternatively, we could have `knex` open the hood on the `pg` `Client` object and figure things out on its own.  But, this will couple `knex` to the current `pg` implementation, which will be problematic in the future.

**Comment by jamesdixon** [comment]
> @jamesdixon this has been around as long as AWS lambda with pg has existed.

@elhigu I'm not sure what you're referring to specifically? There has definitely been a behavior change between Node 8 and Node 10 / 12 even if the underlying internals in Knex/pg have remained the same.

**Comment by jamesdixon** [comment]
@592da I'm not sure why using a Node 8 runtime isn't working for you. It resolved the issue immediately for me. Can you verify that in the AWS console that your functions are definitely set to use your Node 8 runtime?

**Comment by briandamaged** [comment]
Wow -- GitHub has certainly been on the struggle buggy lately!  HTTP 500 responses all over the place!

Anyhoo -- the underlying issue seems to be that both the `knex` and `pg` pool implementations are relying upon Events to detect/evict closed connections.  But, if these Events are not handled in time, then the pools can hand off closed connections.

@kibertoad mentioned that there were changes to the `EventEmitter` class between `node 8` and `node 10`.  I don't know the details of these changes, but they could potentially impact the order in which Events get processed.  If so, then this could also explain why some end-users are not observing the issue in `node 8`.

**Comment by elhigu** [comment]
> @elhigu I'm not sure what you're referring to specifically? There has definitely been a behavior change between Node 8 and Node 10 / 12 even if the underlying internals in Knex/pg have remained the same.

I mean that I remembered that there has been unresolved reports of knex not working properly with AWS lambda before this issue. But now that I searched for them looks like they were about transaction hanging https://github.com/knex/knex/issues/2445 

So I stand corrected. It might not be related to this issue after all.

@briandamaged 
> @elhigu : It has it built-in, but the base implementation still relies upon certain Events being fired. These events attach a __knex__disposed field to the connection object to indicate that the connection has been closed. This field is then checked by the pool before handing off the connection.

Correct. Thats why I suggested that adding delay before query is made could allow node event loop to have time to recognize that connection has been ended / is broken already when it is still in the pool. 

Though I don't know how node's TCP socket recognizes this. If it happens by TCP sockets keealive feature looks like the delay needs to be huge until it would notice that. https://stackoverflow.com/questions/18614176/why-nodejs-keepalive-does-not-seem-to-work-as-expected

So I suppose if this works with Node 8 that there must be also some other way to notice it.

> So, I believe this is the reason that some end-users are still seeing this issue in knex. If the Events are not fired/handled quickly enough, then a connection can slip through the cracks.
> Fortunately, knex also provides a hook so that each Dialect can add its own connection validation logic. For Client_PG, this method currently just returns true. If the pg team adds a public method for checking the state of a Client, then we can update the Client_PG's hook to query this state.
> Alternatively, we could have knex open the hood on the pg Client object and figure things out on its own. But, this will couple knex to the current pg implementation, which will be problematic in the future.

@brianc has been always really helpful with providing necassary APIs to be able to prevent that. If the problem really is recognizable inside pg driver, I'm sure we get some official way to check that :)

I just wish there would be a way to check that out from connection object, without need to create an additional query. Adding dummy query before getting connection from pool would not be suitable default behaviour since it adds too much overhead. Though that could overridden in pool configuration when knex is used with lambda.

Also Tarn.js doesn't currently support async validate functions (I started writing support for it, but it is not complete yet, it needed pretty big changes to make that possible).

**Comment by briandamaged** [comment]
@elhigu : when I said "query", I meant "query a predicate method that returns synchronously".  I'm actually in the process of implementing an experimental branch in `knex` for this.  If the experiment works, then the `pg` team can integrate the concept / expose a public API for it.

(Unfortunately, I'm encountering a minor roadblock at the moment.  It appears that a handful of Postgres unit tests are attempting to connect to the wrong port.  I'm not sure why this fix would suddenly uncover that issue)

**Comment by brianc** [comment]
> If the pg team adds a public method for checking the state of a Client, then we can update the Client_PG's hook to query this state.

I can add this - would you open an issue on [node-postgres](https://github.com/brianc/node-postgres) & outline what you'd want this method to check for?  We can hash out the details there.

**Comment by briandamaged** [comment]
@brianc : Ah -- it's the same thing we're hashing out in this PR:

https://github.com/brianc/node-postgres/pull/2121

In particular, I'm creating an experimental build that will use the `stream.readable` and `stream.writable` properties.  (And possibly a few others).  Once this is ready, I'm hoping that @jamesdixon  can try it out on his Lambda instance and see if the problems persist.

**Comment by brianc** [comment]
> Ah -- it's the same thing we're hashing out in this PR:

oh okay we can use that PR I suppose to continue the discussion. I'll follow up over there.

**Comment by briandamaged** [comment]
@jamesdixon : I've created an experimental Knex build here:

https://github.com/briandamaged/knex/tree/experiment/aws-disconnections

Would it be possible for you to try to recreate the issue using this build?  If it works, then this should help us figure out a more permanent solution.  Thx!

cc: @brianc @elhigu

**Comment by jamesdixon** [comment]
Yes! I should be back at my computer in a Few hours.

🙏

**Comment by jamesdixon** [comment]
@briandamaged I've updated my Lambda with this branch but it may take a while to actually determine if the issue still exists. The error hits randomly.

**Comment by briandamaged** [comment]
@jamesdixon : Sounz good!

Based upon the current discussions, I'm not super-confident that this is experiment will work.  But, we might be pleasantly surprised.  🙏

**Comment by briandamaged** [comment]
@jamesdixon : Howdy!  Any new developments on this front?  Or, is it still too early to tell?

**Comment by elhigu** [comment]
usually this kind of fixes works until feature gets released and then it appears again 😬😅

**Comment by jamesdixon** [comment]
@briandamaged not yet. I need to push some "volume" through the lambda in question in order to see if the issue is triggered. I have to run some testing tomorrow that should do the trick. Apologies for the delay.

If anyone else wants to give it a shot, it would be great to have some additional confirmation on whether it does or does not work 👍

**Comment by jamesdixon** [comment]
@briandamaged sorry for the delay. I ended up pushing this new version to my Lambda job that handles sending emails and unfortunately, the issue popped up within a few minutes 😢

**Comment by briandamaged** [comment]
@jamesdixon - Ah, dang.  Well, I guess that answers _that_ question!

So, I'm starting to think that we might need to take a fundamentally different approach to solving this problem.  It seems that some of `knex`'s current assumptions about connection pooling / retention might be incompatible with Lambda.

In the near-term, you can try moving your pool setup/teardown logic into the Lambda handler, as mentioned in this comment:

https://github.com/brianc/node-postgres/issues/2112#issuecomment-592459198

Presumably, this will limit the lifetime of each pool (and thus its connections) to a single Lambda transaction.

**Comment by Ali-Dalal** [comment]
Hi everyone,

Did anyone try RDS proxy with Knex?

https://aws.amazon.com/blogs/compute/using-amazon-rds-proxy-with-aws-lambda/

Do you think this might help mitigate the issue with `connection terminated unexpectedly`?

**Comment by lastmjs** [comment]
I think I might be running into issues with Knex and the RDS Lambda Proxy. All or most of my database connections are pinned, and I think it's because of prepared statements. Does Knex use prepared statements for all queries by default? Is there any way to turn them off?

**Comment by elhigu** [comment]
Knex creates a new prepared statement on every query. That cannot be turned off. Extended query protocol I suppose depends how have you setup driver configuration. Knex doesn't know anything about it.

**Comment by chrisjhoughton** [comment]
@lastmjs I'm pretty sure I saw this issue a while back and updating to the latest version of `pg` fixed it 👍

**Comment by lastmjs** [comment]
@chrisjhoughton I was already on `8.2.1` of `pg`, now I'm updating `8.3.3`. How long ago do you think you saw this? I'll be testing imminently

**Comment by chrisjhoughton** [comment]
@lastmjs my bad, we're still seeing it. V small issue though.

**Comment by lastmjs** [comment]
@chrisjhoughton You're saying it's a small issue for you? The RDS Proxy documentation says that the extended query protocol causes session pinning...won't all or nearly all parameterized queries use the extended query protocol? If you're having a small issue with this, I'm not sure why I'm having such a large issue.

**Comment by chrisjhoughton** [comment]
We haven't switched our Sequelize app over to RDS proxy yet, but we saw it in another app with Knex (which also uses `pg`) which caused every request to fail. Updating `pg` and `knex` to the latest versions fixed it. Sorry I can't help on the Sequelize front yet - although we're planning on switching to RDS proxy with it soon.

**Comment by lastmjs** [comment]
After digging into this more, it seems not possible to use prepared statements with any kind of transaction pooling on postgres. The two main types of pooling are session pooling and transaction pooling. Transaction pooling seems the most generally performant. Transaction pooling is what RDS Proxy uses. `node-postgres` or `pg` does not seem compatible with any kind of transaction pooling if using prepared statements, and Knex I believe always uses prepared statements. This discussion that I've started is probably better moved over to the pg repo. I'll leave some links here:

* https://github.com/brianc/node-postgres/issues/2266
* https://github.com/brianc/node-postgres/issues/2326

**Comment by rendyfebry** [comment]
Hi everyone, I also experienced this behaviour with Postgres + NLB setup

If I understand correctly this is not only a Lambda problem but just connection lost in general. NLB has this broken behaviour, it will just timeout after being idle for 350s, SILENTLY WITHOUT telling the client. At this point, knex/pg doesn't know the connection already closed, so when the next query come-in knex/pg will try to connect with the previous connection, then realize the connection already broken, and will throw connection terminated unexpectedly error. This [blog](https://medium.com/tenable-techblog/lessons-from-aws-nlb-timeouts-5028a8f65dda) sums up this behaviour clearly.

Steps to reproduce:

1. Put NLB in front of your Postgres server
2. Connect your script to the NLB instead of to the Postgres server directly
3. Make one success request/query, to trigger the pool
4. Wait for 350+ seconds
5. Make request/query again, you will see connection terminated unexpectedly error.

Oh in my case, psql also facing the same problem, which should not be a problem for an interactive tool like psql. But for knex/pg, my expectation is to retry internally when realized the connection already broken.

**Comment by elhigu** [comment]
> But for knex/pg, my expectation is to retry internally when realized the connection already broken.

This cannot work robustly for example because transactions. You should configure your pool to have min: 0 connections and set idleTimeoutMillis to be like 300000ms so that pool will discard connection before server does.

**Comment by rendyfebry** [comment]
> > But for knex/pg, my expectation is to retry internally when realized the connection already broken.
> 
> This cannot work robustly for example because transactions. You should configure your pool to have min: 0 connections and set idleTimeoutMillis to be like 300000ms so that pool will discard connection before server does.

Yeah, I read about the `min: 0` and `idleTimeoutMillis` on another issue page, tried that, but no luck so far, the same errors still happening. So I still trying to figure this out.

**Comment by EmilIvanov** [comment]
@jamesdixon
@rendyfebry 

Try this:
```min: 0```
```idleTimeoutMillis: < whatever your server value for closing idle client connections is```
Trigger an idle connection check on pool at the start of your handler
```knex.client.pool.check()```

**Comment by rendyfebry** [comment]
@EmilIvanov @elhigu 

Forgot to update, yes set `idleTimeoutMillis` < NLB timeout + `min: 0` works on latest Knex.

My problem is, I use very old Knex which still using `generic-pool` instead of `tarn`, and `idleTimeoutMillis` actually belong to `tarn`. So I just update my Knex, and it works.

**Comment by knowroozi** [comment]
We have been trying to debug this same situation with notes from this along with other issues referenced. Not sure if this helps, but we are only seeing the connection timeouts happen when we place the lambdas inside a VPC. The access to the database through the VPC is correct and all requests are successful, but after some amount of time intermittently the `Connection terminated unexpectedly` happens. We tried the bare minimum of 1 AZ in the VPC so only 1 process is spawned, but that process won't hold the connection the same way when it is just placed in a public subnet.

**Comment by maxogden** [comment]
I was able to stop 'Connection terminated unexpectedly' on Lambda Node.js 12.x/Knex 0.19.5/Aurora PostgreSQL by adding these `knex` constructor options:

```
acquireConnectionTimeout: 5000,
pool: {
  min: 0,
  max: 10,
  createTimeoutMillis: 8000,
  acquireTimeoutMillis: 8000,
  idleTimeoutMillis: 8000,
  reapIntervalMillis: 1000,
  createRetryIntervalMillis: 100,
  propagateCreateError: false
}
```

**Comment by elhigu** [comment]
Please don't set `propagateCreateError: false` it will make knex behave badly.

https://github.com/knex/knex/issues/3455#issuecomment-535554401
https://github.com/knex/knex/issues/3447#issuecomment-535577333
https://github.com/knex/knex/issues/3362#issuecomment-513288860
https://github.com/knex/knex/issues/3310#issuecomment-506105627

I'm thinking that maybe we should setup knex to throw an error if someone does that.

**Comment by maxogden** [comment]
@elhigu thanks, I got the advice from this blog post: https://deniapps.com/blog/setup-aws-lambda-to-use-amazon-rds-proxy

**Comment by deniapps** [comment]
@maxogden I am sorry for providing the wrong information. I believe that I meant to set that value to be `true`. I mentioned the settings in many places including [this one](https://github.com/knex/knex/issues/3464#issuecomment-678697726), and they all are set to `true`, except the one you were using. But as @elhigu said, the setting of `progapateCreateError` should not be touched, so I have removed it from all the example config in my blog.

@elhigu, just curious, why not set this value to be read-only in your library if you don't like the user to touch it?

**Comment by elhigu** [comment]
> @elhigu, just curious, why not set this value to be read-only in your library if you don't like the user to touch it?

We just haven't wanted to touch pool configuration that is passed pretty much directly to `tarn.js`. For example one can override create/destroy callbacks with broken ones if one wants. Also there are some cases where actually someone could like to set that parameter `false`, but one would need to know exactly what one is doing in that case.

**Comment by maxogden** [comment]
Ok removing `progapateCreateError` brought the `Connection terminated unexpectedly` errors back in Lambda production. My hacky fix is to do a test query on every lambda invocation

```js
const knexTestQuery = async () => {
  try {
    // do a test query to make sure the knex connection pool is ready to go after lambda unfreeze
    knex('events').select('*').first()
  } catch (err) {
    // log but otherwise ignore errors here
    logger.warn('test query error', err.message)
  }
}
```

I believe the issue is that when lambda goes from frozen to thawed state and then tries to do a query, it has a chance of picking up a bad client from the pg pool (note: the knex instance is only created at lambda function allocation time, but a function may be frozen and thawed multiple times in the lifecycle of a function instance). Doing this test query will catch any bad clients and cause the pool to heal itself before my actual query runs. I have the same issue with `ioredis`, but `ioredis` provides APIs for connecting and disconnecting, so on every lambda invocation I start with `redis.connect()` and at the end I do `redis.disconnect()`. I couldn't figure out how to disconnect and reconnect a knex instance without destroying and re-creating it so I'm doing the hacky test query method instead

**Comment by elhigu** [comment]
> disconnect and reconnect a knex instance without destroying and re-creating

```
await knex.destroy();
knex.init();
```

should drop old connections and reinitialize pool. It is a mystery why `pg` doesn't get any events of broken connections during the sleep. 

We might also add an option to pool to force reaping cycle to run, which could remove all the old connections from pool: https://github.com/Vincit/tarn.js/issues/56

**Comment by ericmaicon** [comment]
any news on this one? I have 3 environments with the same code in AWS...different accounts. Suddenly I started to received it in on of them. Not in the others.

**Comment by kikeflowers** [comment]
> # Environment
> Knex version: 0.20.8
> Database + version: Postgres 9.5.14
> OS: AWS Lambda - Node 12.x
> 
> Recently, I've been seeing upticks in a `connection terminated unexpectedly` error. I can't find a discernible pattern at this point but unlike #3523, it doesn't seem to occur after long periods of activity. I see instances that are minutes apart in some cases.
> 
> I added some logging to the Knex `afterCreate` method and am seeing two errors:
> 
> 1. The `connection terminated unexpectedly` error
> 2. And `Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed` - this seems to happen on the 1st or second invocation after seeing the `connection terminated unexpectedly` error.
> 
> I've tried a number of configurations with Knex pooling:
> 
> 1. `{ pool: { min: 1, max: 1 } }`
> 2. `{ pool: { min: 0  }`
> 3. `{ pool: { min: 1 } }`
> 
> I've also tried setting `{ pool: { idleTimeoutMillis: 300000 } }` as per #3523.
> 
> Additional information:
> 
> * I'm defining my Knex instance outside of my AWS handlers, so the pool should be reused per container.
> * I'm using the following for my `afterCreate` function -- I'm not sure if there's a better way to get at the underlying error?
> 
> ```js
> function afterCreate(connection, done) {
>   connection.on('error', (err) => {
>     console.log('POSTGRES CONNECTION ERROR');
>     console.error(err)
>   });
>   done()
> }
> ```
> 
> Appreciate any help! This is driving me nuts :)
> 
> **UPDATE:** the only change that I've made is migrating the lambda function to Node 12 from Node 8 as AWS is removing support for 8. Unfortunately, it doesn't look like I can go back. My other functions in production are on Node 8 and not seeing any issues 
> 
> **ISSUE FOUND:** for whatever reason, reverting back to Node 8.x fixes the issue. Still not certain on the underlying cause.

I downgraded to node 8.x and it does not worked

**Comment by Warkanlock** [comment]
**This could be related to this issue**
https://github.com/knex/knex/issues/3523#issuecomment-864553289

**Comment by Dakuan** [comment]
very much still an issue

**Comment by sharadpattanshetti** [comment]
This is still an issue for us with backstage using tarnjs pool config and application deployed in K8S env. We tried below config as well but max connections remain 10 only and throw the same error.

{ pool: { min: 0, max: 20}

**Comment by leandro-cam** [comment]
> This is still an issue for us with backstage using tarnjs pool config and application deployed in K8S env. We tried below config as well but max connections remain 10 only and throw the same error.
> 
> { pool: { min: 0, max: 20}

I had problem of loss of database connection due to idleness because of AWS NLB (Network Load Balancer).

I managed to fix it in my Backstage application as follows:

Configuring:

idleTimeoutMillis for 300000 ms

min connections to 0 - min: 0 

This is important because the tarn.js doc says:

"Free resources are destroyed after this many milliseconds. Note that if min > 0, some resources may be kept alive for longer. To reliably destroy all idle resources, set min to 0"

https://github.com/Vincit/tarn.js#:~:text=//%20minimum%20size%0A%20%20min,propagateCreateError%3A%20false

And, lastly, setting reapIntervalMillis to 30000 ms (without this I continued to look for canceled connections).

```
pool:
        min: 0
        maximum: 30
        acquireTimeoutMillis: 60000 # Acquire promises are rejected after this many milliseconds, if a resource cannot be acquired
        createTimeoutMillis: 30000 # Create operations are canceled after this many milliseconds, if a resource cannot be acquired
        idleTimeoutMillis: 300000 # Free resources are destroyed after this many milliseconds. Note that if min > 0, some resources may be kept alive for longer. To reliably destroy all idle resources, set min to 0
        reapIntervalMillis: 30000 # How often to check for idle resources to destroy
```

**Comment by mercmobily** [comment]
I am closing this, since it's basically identical to #3523. To be honest, it was _really_ difficult to choose the main ticket, since they BOTH have very valuable information and comments. I am in the process of tidying up the issue list, hence the closing of this one. I encourage everybody to jump over #3523 if you are still interested in this problem, and if you have any hints.
